go/6_auth-kubernetes.go

```diff
@@ -0,0 +1,92 @@
+package main
+
+import (
+	"fmt"
+	"os"
+
+	vault "github.com/hashicorp/vault/api"
+)
+
+// Fetches a key-value secret (kv-v2) after authenticating to Vault with a Kubernetes service account.
+//
+// As the client, all we need to do is pass along the JWT token representing our application's Kubernetes Service Account in our login request to Vault.
+// This token is automatically mounted to your application's container by Kubernetes. Read more at https://www.vaultproject.io/docs/auth/kubernetes
+//
+// SETUP NOTES: If an operator has not already set up Kubernetes auth in Vault for you, then you must also first configure the Vault server with its own Service Account token to be able to communicate with the Kubernetes API
+// so it can verify that the client's service-account token is valid. The service account that will be performing that verification needs the ClusterRole system:auth-delegator.
+//
+//    export TOKEN_REVIEW_JWT=$(kubectl get secret $TOKEN_REVIEWER_SECRET --output='go-template={{ .data.token }}' | base64 --decode)
+//    export KUBE_HOST=$(kubectl config view --raw --minify --flatten --output='jsonpath={.clusters[].cluster.server}')
+//    kubectl config view --raw --minify --flatten --output='jsonpath={.clusters[].cluster.certificate-authority-data}' | base64 --decode > path/to/kube_ca_cert
+//
+//    vault write auth/kubernetes/config \
+//  	token_reviewer_jwt=${TOKEN_REVIEW_JWT} \
+//      kubernetes_host=${KUBE_HOST} \
+//      kubernetes_ca_cert=@path/to/kube_ca_cert \
+//      issuer="kubernetes/serviceaccount"
+//
+// The "issuer" field only needs to be set when running Kubernetes version 1.21 or above:
+// https://www.vaultproject.io/docs/platform/k8s/csi#setting-issuer-for-kubernetes-authentication.
+//
+// Finally, make sure to create a role in Vault bound to your pod's service account:
+//
+// 	vault write auth/kubernetes/role/dev-role-k8s \
+//     	policies="dev-policy" \
+//     	bound_service_account_names="my-app" \
+//		bound_service_account_namespaces="default"
+func getSecretWithKubernetesAuth() (string, error) {
+	// If set, the VAULT_ADDR environment variable will be the address that your pod uses to communicate with Vault.
+	// For an application running in Kubernetes accessing a Vault server outside of Kubernetes, you will need to set VAULT_ADDR in your pod spec to
+	// point at your Kubernetes cluster's gateway address. See this tutorial for more information: https://learn.hashicorp.com/tutorials/vault/kubernetes-external-vault
+	config := vault.DefaultConfig() // modify for more granular configuration
```

Comment:
Is there anything in the DefaultConfig that isn't obvious and should be noted in the comments?

Reply:
(Unrelated to the PR) I'm also curious if there are extensions for Go for you to view referenced code, so you don't have to pull up the github repo to figure out what some of the utilities do?

Reply:
I started the first examples initially with just initializing a Config struct, but Nick pointed out that DefaultConfig is usually what the team prefers for demos/tutorials. The only thing that's unusual with Kubernetes in particular is the thing I mentioned in the comment above this line (the cluster gateway bit).

As for the extensions, I assume you mean in an editor? Yes, in VSCode I have the official Go extension installed, and that allows me to option-click on any function and it will transport me over to that function's definition. (right-click -> Go To Definition works too). I imagine other editors like GoLand have similar functionality.

Reply:
Sounds good!